File: python/sort/merge_sort.py
# 归并排序
from typing import List


def merge_sort(data: List[int]) -> None:
    merge_sort_helper(data, 0, len(data) - 1)


def merge_sort_helper(nums: List[int], p, r):
    if p >= r:
        return
    q = (p + r) // 2
    merge_sort_helper(nums, p, q)
    merge_sort_helper(nums, q + 1, r)
    merge(nums, p, q, r)


def merge(nums: List[int], p, q, r):
    temp = []
    i, j = p, q + 1
    while i <= q and j <= r:
        if nums[i] < nums[j]:
            temp.append(nums[i])
            i += 1
        else:
            temp.append(nums[j])
            j += 1
    if i <= q:
        temp[len(temp):] = nums[i:q + 1]
    else:
        temp[len(temp):] = nums[j:r + 1]
    nums[p:r + 1] = temp[:]

File: go/sort/merge_sort.go
package sort

//归并排序

func MergeSort(data []int) {
	mergeSortHelper(data, 0, len(data)-1)
}

func mergeSortHelper(data []int, lo, hi int) {
	if lo < hi {
		mid := lo + (hi-lo)/2
		mergeSortHelper(data, lo, mid)
		mergeSortHelper(data, mid+1, hi)
		merge(data, lo, mid, hi)
	}
}

func merge(data []int, lo, mid, hi int) {
	temp := make([]int, hi-lo+1)
	i, j, k := lo, mid+1, 0
	for i <= mid && j <= hi {
		if data[i] < data[j] {
			temp[k] = data[i]
			i++
		} else {
			temp[k] = data[j]
			j++
		}
		k++
	}
	copy(temp[k:], data[i:mid+1])
	copy(temp[k:], data[j:hi+1])
	copy(data[lo:hi+1], temp[:])
}
